src/string_search.h

```diff
@@ -274,111 +229,133 @@ inline uint8_t GetHighestValueByte(uint16_t character) {
 inline uint8_t GetHighestValueByte(uint8_t character) { return character; }
 
 
-template <typename PatternChar, typename SubjectChar>
-inline size_t FindFirstCharacter(Vector<const PatternChar> pattern,
-                              Vector<const SubjectChar> subject, size_t index) {
-  const PatternChar pattern_first_char = pattern[0];
+// Searches for a byte value in a memory buffer, back to front
+// Uses memrchr(3) on systems which support it, for speed
+// Falls back to a vanilla for loop on systems which don't
+inline const void* MemrchrFill(const void* haystack, uint8_t needle,
+                               size_t haystack_len) {
+#ifdef _GNU_SOURCE
+  return memrchr(haystack, needle, haystack_len);
+#else
+  const uint8_t *haystack8 = reinterpret_cast<const uint8_t*>(haystack);
+  for (size_t i = haystack_len - 1; i != (size_t)-1; i--) {
+    if (haystack8[i] == needle) {
+      return haystack8 + i;
+    }
+  }
+  return nullptr;
+#endif
+}
+
+
+// Finds the first occurance of *two-byte* character pattern[0]
+// in the string `subject`. Does not check that the whole pattern matches.
+template <typename Char>
+inline size_t FindFirstCharacter(Vector<const Char> pattern,
+                              Vector<const Char> subject, size_t index) {
+  const Char pattern_first_char = pattern[0];
   const size_t max_n = (subject.length() - pattern.length() + 1);
 
+  // For speed, search for the more `rare` of the two bytes in pattern[0]
+  // using memchr / memrchr (which are much faster than a simple for loop)
   const uint8_t search_byte = GetHighestValueByte(pattern_first_char);
-  const SubjectChar search_char = static_cast<SubjectChar>(pattern_first_char);
   size_t pos = index;
   do {
-    const SubjectChar* char_pos = reinterpret_cast<const SubjectChar*>(
-        memchr(subject.start() + pos, search_byte,
-               (max_n - pos) * sizeof(SubjectChar)));
+    size_t bytes_to_search;
+    const void* void_pos;
+    if (subject.forward()) {
+      bytes_to_search = (max_n - pos) * sizeof(Char);
+      void_pos = memchr(subject.start() + pos, search_byte, bytes_to_search);
+    } else {
+      bytes_to_search = (subject.length() - pos) * sizeof(Char);
+      void_pos = MemrchrFill(subject.start(), search_byte, bytes_to_search);
+    }
+    const Char* char_pos = reinterpret_cast<const Char*>(void_pos);
     if (char_pos == nullptr)
       return subject.length();
-    char_pos = AlignDown(char_pos, sizeof(SubjectChar));
-    pos = static_cast<size_t>(char_pos - subject.start());
-    if (subject[pos] == search_char)
+
+    // Then, for each match, verify that the full two bytes match pattern[0]
+    char_pos = AlignDown(char_pos, sizeof(Char));
+    size_t raw_pos = static_cast<size_t>(char_pos - subject.start());
```

Comment:
No `char_pos >= subject.start()`, regardless of whether we're searching forward or backward.

We're looking for `pattern` in `subject`. Every `char_pos` value is a potential match and should always be within the bounds of `subject`